launchpad/src/launchpad.rs

```diff
@@ -3,324 +3,340 @@
-        self.require_stage(LaunchStage::Claim)?;
+    fn claim_launchpad_tokens(
+        &self,
+        #[var_args] opt_nr_tickets_to_redeem: OptionalArg<usize>,
+    ) -> SCResult<()> {
+        require!(self.winners_selected().get(), "Winners not selected yet");
+        self.require_claim_period()?;
 
         let caller = self.blockchain().get_caller();
-        require!(
-            !self.blacklist().contains(&caller),
-            "You have been put into the blacklist and may not claim tokens"
-        );
+        require!(self.has_user_claimed(&caller), "Already claimed");
 
         let (first_ticket_id, last_ticket_id) = self.try_get_ticket_range(&caller)?;
-        let mut nr_redeemed_tickets = 0u32;
+        let total_tickets = last_ticket_id - first_ticket_id + 1;
+        let mut nr_redeemable_tickets = 0;
 
         for ticket_id in first_ticket_id..=last_ticket_id {
             let ticket_status = self.ticket_status(ticket_id).get();
-            if !ticket_status.is_confirmed(None) {
-                continue;
+            if ticket_status == WINNING_TICKET {
+                self.ticket_status(ticket_id).clear();
+
+                nr_redeemable_tickets += 1;
             }
 
-            self.ticket_status(ticket_id).set(&TicketStatus::Redeemed);
+            self.ticket_pos_to_id(ticket_id).clear();
+        }
+
+        let nr_tickets_to_redeem = opt_nr_tickets_to_redeem
+            .into_option()
+            .unwrap_or(nr_redeemable_tickets);
+
+        require!(
+            nr_tickets_to_redeem <= nr_redeemable_tickets,
+            "Trying to redeem too many tickets"
+        );
 
-            nr_redeemed_tickets += 1;
+        self.confirmed_all_tickets(&caller).clear();
+        self.ticket_range_for_address(&caller).clear();
+        self.ticket_batch(first_ticket_id).clear();
+
+        if nr_tickets_to_redeem > 0 {
+            let ticket_price = self.ticket_price().get();
+            let redeemed_ticket_cost = &Self::BigUint::from(nr_tickets_to_redeem) * &ticket_price;
+            self.claimable_ticket_payment()
+                .update(|claimable_ticket_payment| {
+                    *claimable_ticket_payment += redeemed_ticket_cost
+                });
         }
 
-        require!(nr_redeemed_tickets > 0, "No tickets to redeem");
+        let nr_leftover_tickets = nr_redeemable_tickets - nr_tickets_to_redeem;
+        if nr_leftover_tickets > 0 {
+            let launchpad_tokens_per_winning_ticket =
+                self.launchpad_tokens_per_winning_ticket().get();
+            let leftover_amount =
+                Self::BigUint::from(nr_leftover_tickets) * launchpad_tokens_per_winning_ticket;
 
-        let launchpad_token_id = self.launchpad_token_id().get();
-        let tokens_per_winning_ticket = self.launchpad_tokens_per_winning_ticket().get();
-        let amount_to_send = Self::BigUint::from(nr_redeemed_tickets) * tokens_per_winning_ticket;
+            self.leftover_launchpad_tokens()
```

Comment:
in what kind of scenario do we have to update the leftover tickets ? I do not think this code is necessary.